File: assign1/ProdCons3.java
import java.util.LinkedList;
import java.util.Queue;

class ProdCons3 extends Thread {
  static final int BUFSIZE = 20; 
  static final int NUMITEMS = 100; 
  static int cc[], pc[];
  static Object lck = new Object();
  static Queue<Integer> q = new LinkedList<>();

  public static void main(String[] args) throws InterruptedException {
    int num_p, num_c = 0;
    int tot = args.length;
    if(tot == 0){ 
      num_c = nump = 1;
    } else if (tot == 1) { 
      num_c = Integer.parseInt(args[0]);
      num_p = 1;
    } else {
      num_c = Integer.parseInt(args[0]);
      num_p = Integer.parseInt(args[1]);
    }

    final Prod p = new Prod();
    final Con[] cons = new Con[k];
    ret = new int [k];
    
    
    Thread prod = new Thread() {
    };
    prod.start();

    for(int i = 0; i < k; ++i){
      cons[i] = new Con(i);
    }
    for(int i = 0; i < k; ++i) {
      cons[i].start();
    }
    for(int i = 0; i < k; ++i) {  
      cons[i].join();
    }
    prod.join();

    System.out.print("Consumer Stats: [");
    for(int i = 0; i < k; ++i) {  
      System.out.print(ret[i] + ",");
      tot += ret[i];
    }
    System.out.println("] total = " + tot);
    
        
    System.out.println("Main: all done! So leave me alone!");
  }
  public static class Prod {
    Prod() {}

    private void producer() throws InterruptedException{ 
      int i = 0;
      System.out.println("Producer is starting...");
      while(i < NUMITEMS){
        ++i;
        synchronized(lck) {
          while(q.size() == BUFSIZE) {
            lck.notify();
            lck.wait();
          }
          System.out.println("Producer added " + i + " (qsz: " + q.size() + ")");
          q.add(i);
          lck.notify();
        }
      }
      synchronized(lck) {
        for(i = 0; i < NUMITEMS; ++i)
          q.add(-1);
          lck.notifyAll();
      }
      System.out.println("Producer is ending...");
    }

    public void run() 
    {
      try {
        int i = this.producer();
        pc[this.nme] = i;
      } 
      catch (InterruptedException e) {
        e.printStackTrace();
      }
    }
  }

  public static class Con extends Thread{
    private int nme = 0;
    Con(int tid) {
      nme = tid;
    }

    private int consumer()  throws InterruptedException{ 
      System.out.println("Consumer[" + currentThread().getName() + "] is starting...");
      int num = 0;
      int i = 0;

      while(num != -1){
        synchronized(lck) {
          while(q.size() == 0) {
            lck.wait();
          }
          num = q.remove();
          if( num != -1) {
            System.out.println("Consumer[" + currentThread().getName() 
                + "] consumed " + num + " (qsz: " + q.size() + ")");
            ++i;
          }
          lck.notify();
        }
      }

      System.out.println("Consumer[" + currentThread().getName() + "] is ending...");
      return i;
    }

    public void run(){
      try {
        int i = this.consumer();
        cc[this.nme] = i;
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }
  }
}

File: assign1/prodcons3.cpp
#include <iostream> 
#include <string> 
#include <thread> 
#include <mutex> 
#include <condition_variable> 
#include "queue.h"
using namespace std; 

mutex mtx;                  //mutex for the conditional
condition_variable con_cv;  //consumer's cv, con wait on empty Q
condition_variable prod_cv; //producer's cv, prod wait on full Q
int BUFFSIZE = 20;
int NUMITEMS = 100; 
int * conRes;               //results array for the consumer
int * prodRes;              //results array for the producer
int numProds;               //number of producers
Queue pcq(BUFFSIZE);

/* Called from main. It waits on the producer's conditional
 * variable when the queue is full otherwise adds a number
 * to the queue.
*/
void producer(int tid) {
  int core = sched_getcpu();
  int range = NUMITEMS/numProds;
  int start = (range * tid) + 1;
  int stop = start + range - 1;
  int i, c = 0;
  if(tid == numProds - 1)
    stop = NUMITEMS;

  cout << "Producer[" + to_string(tid) + "] starting on core " + to_string(core)
    + " for segment [" + to_string(start) + ".." + to_string(stop) + "]\n";

  for(i = start; i <= stop; i++){
    unique_lock<mutex> lck(mtx);
    while(pcq.isFull()) 
    {
      con_cv.notify_all();
      prod_cv.wait(lck);
    } 
    pcq.add(i);
    con_cv.notify_one();
    lck.unlock();
    ++c;
    cout << "Producer[" + to_string(tid) + "] added " + to_string(i) 
      + " (qsz: " + to_string(pcq.size()) + ")\n"; 
  }
  prodRes[tid] = c;                             //records the count it added
  cout << "Producer[" + to_string(tid) + "] ending\n";
}
  
/* Called from main. It waits on the consumer's conditional
 * variable when the queue is empty otherwise removes a number
 * from the queue.
*/
void consumer(int tid) {
  int r, c = 0;
  int core = sched_getcpu();

  cout << "Consumer[" + to_string(tid) + "] starting on core " + to_string(core) + "\n";

  while(r != -1) {
    unique_lock<mutex> lck(mtx);
    while(pcq.isEmpty()) 
    {
      prod_cv.notify_all();
      con_cv.wait(lck);
    } 
    r = pcq.remove();
    prod_cv.notify_one();
    lck.unlock();
    if (r != -1) {
      ++c;
      cout << "Consumer[" + to_string(tid) + "] removed " + to_string(r) 
        + " (qsz: " + to_string(pcq.size()) + ")\n"; 
    }
  }

  prod_cv.notify_one();
  conRes[tid] = c;
  cout << "Consumer[" + to_string(tid) + "] ending\n";
}

int main(int argc, char **argv) {
  int numCons, tot = 0;

  if(argv[1]) numCons = atoi(argv[1]);   //gets the number of
  else numCons = 1;                      //consumers

  if(argv[2]) numProds = atoi(argv[2]);  //gets the number of
  else numProds = 1;                     //producers

  conRes = new int [numCons];            //inits consumer result
  prodRes = new int [numProds];          //inits prod results

  thread pthd[numProds], cthd[numCons];

  cout << "Prodcons3 with "+ to_string(numCons) + " consumers, "+ to_string(numProds) + " producers\n";

  for (int k = 0; k < numProds; k++)     //start ptoducers
    pthd[k] = thread(producer, k);
  for (int k = 0; k < numCons; k++)      //start consumers
    cthd[k] = thread(consumer, k);
  for (int k = 0; k < numProds; k++)     //join producers     
    pthd[k].join();

  for (int k = 0; k < numCons; ++k){     //add -1 to the queue
    unique_lock<mutex> lck(mtx);         //stop the consumers
    while(pcq.isFull()) 
    {
      con_cv.notify_all();
      prod_cv.wait(lck);
    } 
    pcq.add(-1);
    con_cv.notify_one();
    lck.unlock();
  }
  for (int k = 0; k < numCons; k++)     //join consumers
    cthd[k].join();

  cout << "Main added " + to_string(numCons) + " ENDSIGNs (qsz: " + to_string(pcq.size()) + ")\n";

  cout << "Producer stats: [";
  for(int i = 0; i < numProds; ++i) {
    cout << to_string(prodRes[i]) + ",";
    tot += prodRes[i];
  }
  cout << "] = " + to_string(tot) + "\n";

  tot = 0;                               //resets the total count

  cout << "Consumer stats: [";
  for(int i = 0; i < numCons; ++i) {
    cout << to_string(conRes[i]) + ",";
    tot += conRes[i];
  }
  cout << "] = " + to_string(tot) + "\n";
  cout << "Main: all done! Now stop bothering me!\n";
  delete(conRes);
  delete(prodRes);
}  
